### PR TITLE
Add Lichess provider

### PR DESCRIPTION
**Pull request by tanczosm**
- Added Lichess provider and unit test
- Update README.md
- Added documentation

This provider is one of the world's largest chess sites.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -71,6 +71,7 @@ We would love it if you could help contributing to this repository.
 * [Martin Costello](https://github.com/martincostello)
 * [Maxime Roussin-Bélanger](https://github.com/Lorac)
 * [Michael Knowles](https://github.com/mjknowles)
+* [Michael Tanczos](https://github.com/tanczosm)
 * [Patrick Westerhoff](https://github.com/poke)
```
Comment (martincostello) [inline README.md:75]: The new provider needs adding to the table of providers. There's a commented-out final row you can copy to add Lichess to the table.

File: docs/lichess.md [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+## Required Additional Settings
+
+_None._
+
+## Optional Settings
+
+| Property Name | Property Type | Description | Default Value |
+|:--|:--|:--|:--|
+| `Scope` | `string` | The scope you want to use on lichess, see [documentation](https://lichess.org/api#section/Authentication "List of possible scopes"). |  |
```
Comment (martincostello) [inline docs/lichess.md]:
```suggestion
| `Scope` | `ICollection<string>` | The scope(s) you want to use on Lichess, see [documentation](https://lichess.org/api#section/Authentication "List of possible scopes"). |  |
```

File: src/AspNet.Security.OAuth.Lichess/LiChessAuthenticationConstants.cs [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+ * See https://github.com/aspnet-contrib/AspNet.Security.OAuth.Providers
+ * for more information concerning the license and the contributors participating to this project.
+ */
+
+namespace AspNet.Security.OAuth.Lichess
+{
+    /// <summary>
+    /// Contains constants specific to the <see cref="LichessAuthenticationHandler"/>.
+    /// </summary>
+    public static class LiChessAuthenticationConstants
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.Lichess/LiChessAuthenticationConstants.cs]:
```suggestion
    public static class LichessAuthenticationConstants
```

File: src/AspNet.Security.OAuth.Lichess/LichessAuthenticationHandler.cs [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+            context.RunClaimActions();
+
+            // Now retrieve email address if scope is added
+            if (!string.IsNullOrEmpty(Options.UserEmailsEndpoint) &&
+                !identity.HasClaim(claim => claim.Type == ClaimTypes.Email) &&
+                Options.Scope.Contains(LiChessAuthenticationConstants.Scopes.EmailRead))
+            {
+                using var payloadEmail = await RequestUserInformation(Options.UserEmailsEndpoint, tokens.AccessToken, "email address");
+
+                context.RunClaimActions(payloadEmail.RootElement);
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.Lichess/LichessAuthenticationHandler.cs]:
```suggestion
                using var emailPayload = await RequestUserInformation(Options.UserEmailsEndpoint, tokens.AccessToken, "email address");

                context.RunClaimActions(emailPayload.RootElement);
```

File: src/AspNet.Security.OAuth.Lichess/LichessAuthenticationHandler.cs [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+            }
+
+            await Options.Events.CreatingTicket(context);
+            return new AuthenticationTicket(context.Principal!, context.Properties, Scheme.Name);
+        }
+
+        /// <summary>
+        /// Perform a backchannel request to obtain user information at a specified endpoint
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.Lichess/LichessAuthenticationHandler.cs]:
```suggestion
        /// Performs a backchannel request to obtain user information from the specified endpoint
```

File: src/AspNet.Security.OAuth.Lichess/LichessAuthenticationHandler.cs [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+        /// </summary>
+        /// <param name="endpoint">Endpoint to return information from</param>
+        /// <param name="accessToken">Bearer token for authentication</param>
+        /// <param name="requestInformationType">Descriptive text of type of information request in event of exception</param>
+        /// <returns>Parsed Json Document response from endpoint</returns>
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.Lichess/LichessAuthenticationHandler.cs]:
```suggestion
        /// <returns>Parsed JSON document response from endpoint</returns>
```

File: src/AspNet.Security.OAuth.Lichess/LichessAuthenticationHandler.cs [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+        private async Task<JsonDocument> RequestUserInformation(string endpoint, string accessToken, string requestInformationType)
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.Lichess/LichessAuthenticationHandler.cs]:
```suggestion
        private async Task<JsonDocument> RequestUserInformationAsync(string endpoint, string accessToken, string requestInformationType)
```

File: src/AspNet.Security.OAuth.Lichess/LichessAuthenticationHandler.cs [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+            request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", accessToken);
+
+            using var response = await Backchannel.SendAsync(request, HttpCompletionOption.ResponseHeadersRead, Context.RequestAborted);
+            if (!response.IsSuccessStatusCode)
+            {
+                Logger.LogError($"An error occurred while retrieving the {requestInformationType}: the remote server " +
+                                "returned a {Status} response with the following payload: {Headers} {Body}.",
+                    /* Status: */ response.StatusCode,
+                    /* Headers: */ response.Headers.ToString(),
+                    /* Body: */ await response.Content.ReadAsStringAsync());
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.Lichess/LichessAuthenticationHandler.cs]:
```suggestion
                                /* Status: */ response.StatusCode,
                                /* Headers: */ response.Headers.ToString(),
                                /* Body: */ await response.Content.ReadAsStringAsync());
```

**Review by martincostello — COMMENTED** [review]
Just a few minor comments.

**Comment by tanczosm** [comment]
All suggestions have been merged.  Thanks for catching those.

**Comment by martincostello** [comment]
Thanks for you contribution @tanczosm! 🥳

The provider will be released to NuGet.org within the next week or so once I've got some admin tasks on the repository completed.

In the meantime, you can consume a prerelease version of the provider from our MyGet feed once the CI is completed: https://www.myget.org/gallery/aspnet-contrib

**Comment by martincostello** [comment]
Thanks again for your contribution @tanczosm - the provider is now available from NuGet.org as versions 3.1.7 and 5.0.1.

**Pull request merged**